### ISSUE TITLE
Optimizing Leave One Out CV with Talos

### ISSUE THREAD

**Issue by margaro95** [autonomio/talos#15]
Hi!
I have a very little dataset consisting of 59 images of medical data.
I am classifying them with a Keras model of six layers. With such a little dataset, I have to use Leave One Out cross validation, being my objective function the accuracy obtained not on one run of the cross validation but *the mean* of all the 59 runs of the cv (I train 58 images, I test the first image; and so on with the second, third... 59th image).
Is it possible to optimize my model with Talos when I'm in this situation?
Thank you!

**Comment by mikkokotila** [comment]
Do I understand correct that what you do is train the data with the 58 always leaving one out, and then test with the one image that left, and repeat that process total of 59 times always starting new and then in the end take the result e.g. 48 out of 59. Is that correct?

**Comment by margaro95** [comment]
Absolutely correct.
I always train 58 images and test the image that in that run is not on the train set. I do that 59 times (one for each image) and my accuracy would be the mean of accuracies

**Comment by mikkokotila** [comment]
Thanks for clarifying. How do you manage the process currently? Can you share the code that you use for running the experiment. This is not a use-case that had been considered out-of-the-box but it might be very simple to do it, happy to give it a try if you share your code.

**Comment by margaro95** [comment]
The code is the following. Just a normal keras routine. The optimizer should completely run the while loop for every point defined in the grid by talos.

```
from keras.preprocessing.image import ImageDataGenerator
from keras.models import Sequential
from keras.layers import Conv2D, MaxPooling2D, ZeroPadding2D
from keras.layers import Activation, Dropout, Flatten, Dense
from keras import backend as K
from load_alzheimer import load_alzheimer_data

# dimensions of our images.
img_rows, img_cols = 90, 90

train_data_dir = 'data/train'
validation_data_dir = 'data/validation'
nb_train_samples = 58
nb_validation_samples = 1
epochs = 1
batch_size = 5

if K.image_data_format() == 'channels_first':
    input_shape = (1, img_rows, img_cols)
else:
    input_shape = (img_rows, img_cols, 1)

count = 0
J = 0

while count < 59:
    X_train, Y_train, X_valid, Y_valid = load_alzheimer_data(img_rows, img_cols, count)
    
    model = Sequential()
    model.add(ZeroPadding2D(2, input_shape=input_shape))
    model.add(Conv2D(64, (5, 5)))
    model.add(Activation('relu'))
    model.add(MaxPooling2D(pool_size=(2, 2)))

    model.add(Flatten())
    model.add(Dense(64))
    model.add(Activation('relu'))
    model.add(Dropout(0.5))
    model.add(Dense(1))
    model.add(Activation('sigmoid'))

    model.compile(loss='binary_crossentropy',
                  optimizer='rmsprop',
                  metrics=['accuracy'])
    history = model.fit(
        X_train, Y_train,
        batch_size=batch_size,
        epochs=epochs,
        validation_data=(X_valid, Y_valid))
    J += history.history['val_acc'][0]
    count += 1
    print(count)
#model.save_weights('first_try.h5')
print('---------------')
print(J / 59)
```

**Comment by mikkokotila** [comment]
Sorry for the delay in getting back to you on this. It seems that this is not going to work without some modifications. I'm not sure yet what those modifications would be (as I don't want to change the main line code for this) so I'm going think out load here a little first. 

It seems that the issue is in the way a typical model is measured; one model permutation yields one accuracy. Here the case is that each observation yields one accuracy, and then we take average of that in the end. Typically we would have one permutation processed, get the accuracy and then move on to the next permutation, and then in the end pick the best one/s. Here we'd have the case where we'd need to choose if we want to a) repeat the process as you have it in your code, but do it once per permutation, or if we wanted to go through permutations sample by sample. That seems to be the main question here i.e. to do this sample-by-sample and then in the end stitch it together, or do it model-by-model. Sample-by-sample is closer to the current working, but model-by-model makes the accuracy part simpler to do it seems. But because the accuracy part is trivial and does not affect the code much, it seems that sample-by-sample is the preferred approach (even though it seems counter intuitive initially). Does it make sense? 

I'm sorry that at this point it's not going to immediately solve your problem, but it seems worthwhile to figure out the best way to do this, and have it as a feature.

**Comment by mikkokotila** [comment]
...maybe it will be possible to do this in a way where the only thing that changes is a parameter in the Reporting() where the .csv log is processed in a way that makes sense to this use case.

**Comment by margaro95** [comment]
(Sorry for answering this late too)
Definitely, to go through permutations sample by sample is by far the easiest way to implement this feature without changing the main structure in the code.

Everything comes at a price. In the end, we would end up with 59 models most definitely overfitted to each one of its alternating validation image. We could trust our 59 models, show them our test image (the 60th image), get its probas and take the mean. We could blindly trust that naive assumption.

We could also do something more accurate and proffesional. Once we get all those 59 overfitted models, we could train another meta-classifier (LogisticRegressor) that would be trained on the probas that every image gets from all the 58 other overfitted classifiers. The 60th image (test data) would be transformed to that same standard and be shown to that meta-classifier.

But in both these solutions we just end up with a somewhat vaguely accurate accuracy and never get a configuration of parameters from our CNN that we can say is near to optimal.

We are still far from a solution but at least we got something!

**Comment by mikkokotila** [comment]
Sorry for not coming back to this. It's a little bit exotic request, and definitely something would be great to support eventually, but to be frank, it's unlikely to happen unless you would like to take a stab at implementing it (or someone else).

**Comment by mikkokotila** [comment]
Closing this for now for being stale.
